.chronus/changes/add-scalar-type-2024-5-14-13-23-7.md

```diff
@@ -0,0 +1,49 @@
+---
+# Change versionKind to one of: internal, fix, dependencies, feature, deprecation, breaking
+changeKind: breaking
+packages:
+  - "@azure-tools/typespec-client-generator-core"
+---
+
+## Changes
+
+1. remove the following types from `SdkBuiltInType`:
+   - `uuid` (azure type)
+   - `ipV4Address` (azure type)
+   - `ipV6Address` (azure type)
+   - `eTag` (azure type)
+   - `armId` (azure type)
+   - `azureLocation` (azure type)
+   - `password` (removed)
+   - `guid` (removed)
+   - `uri` (removed)
+   - `ipAddress` (removed)
+2. `@format` can no longer change a string type to above azure types (`uuid`, `eTag`, etc).
+3. add `name`, `crossLanguageDefinitionId` and `baseType` to `SdkBuiltInType`, `SdkDatetimeType` and `SdkDurationType`.
+4. now scalars defined using `scalar` keyword will be parsed into either `SdkBuiltInType`, `SdkDatetimeType` or `SdkDurationType` depending on the base type, with its `name` and `crossLanguageDefinitionId`. `@encode` will be added to the scalar type, and will not propagate to its base type.
+
+## Migration guides for emitters
+
+The major breaking change introduced by this change is the removal of azure related kinds for `SdkBuiltInType`. Therefore for code such as:
+```typescript
+if (type.kind === "azureLocation")
+{
+  // do something for azure-location
+}
+```
+should be changed to the following if the emitter is doing something special for azure-location:
+```typescript
+if (type.kind === "string" && type.crossLanguageDefinitionId === "Azure.Core.azureLocation")
```

Comment:
Does dev need to visit the `baseType` in some case?

e.g. if tsp

```ts
scalar myuuid extends Azure.Core.uuid;

model Resource {
  resourceId: myuuid;
}
```

I assume we still treat `myuuid` as UUID?

Reply:
I think we should have an offline discussion for this topic with Chenjie, I have disagreement with him on this on a particular case for encoding.
But in general, the best practice now looks like:
```
function getType(type: SdkType): YourType
{
	if (type.kind === "string")
	{
		// recursively find if any level we have a concerned definition id
		const result = findForSpecial(type);
		if (result) return result;
		return usualString;
	}
	// omit others
}
function findForSpecial(type: SdkBuiltInType & { kind: "string" }): YourType | undefined {
	if (type.crossLanguageDefinitionId === "Something") return something;
	
	if (type.baseType) return findForSpecial(type.baseType);
	
	return undefined;
}
```
I admit this looks weird. Therefore i think we might need an offline discussion for this.

Reply:
this part has been included in the migration guide in the changelog